#140648421 Create a REST endpoint for user registration

giddygitau opened the pull request:
**What does this PR do?**

Have an API endpoint for user registration.

**Description of the task to be completed.**

It should be possible to create new users via the API. External apps might need to register accounts on behalf of their users without having to send them over to the wger instance.

**What are the relevant pivotal tracker stories?**
#140648421

Commit:
Merge pull request #3 from andela/develop

Fix PEP errors

swanjala committed:
Ch setup slack integrations 140648413 (#2)

* [Chore #140648413] Add slack notifications ID

* [Chore #140648413] Add Travis CI badge to README.rst

* [Chore #140648413] Add Coveralls badge to README.rst

* [Chore 140648413] Omit files that are not being tested

* [Chore 140648413] Add the coveralls repository token
Remove python 3.4 and sqlite builds

Commit:
Modify coverall (#5) (#6)

* [Chore #140648413] Add slack notifications ID

* [Chore #140648413] Add Travis CI badge to README.rst

* [Chore #140648413] Add Coveralls badge to README.rst

* [Chore 140648413] Omit files that are not being tested

* [Chore 140648413] Add the coveralls repository token
Remove python 3.4 and sqlite builds
Adding the coveralls installation command

giddygitau committed:
[Feature #140648421] Adds the url to the viewset that contains user creation logic

giddygitau committed:
[Feature #140648421] Creates the viewset that contains logic to create new users

giddygitau committed:
[Feature #140648421] Adds the serializer that deals with user registrations

giddygitau committed:
[Feature #140648421] Adds method to register users, renames self to request

giddygitau committed:
[Feature #140648421] Removes method filters from serializer

giddygitau committed:
[Feature #140648421] Adds test case for user registration

giddygitau committed:
[Feature #140648421] Adds method filter
http_method_names property is a list that contains only the allowable methods in this route

giddygitau committed:
[Feature #140648421] Removes unnecessary User import

giddygitau committed:
[Feature #140648421] Modifies test to use the base test class

egichuri commented on wger/core/api/views.py:
```diff
@@ -39,6 +41,24 @@
+            email=validated_data['email'],
+        )
+        user.set_password(validated_data['password'])
+        user.save()
```
Should the user profile be created when a user registers?

egichuri replied:
If so remember to test that first.

egichuri commented on wger/urls.py:
```diff
@@ -98,6 +98,7 @@
 
 # Core app
 router.register(r'userprofile', core_api_views.UserProfileViewSet, base_name='userprofile')
+router.register(r'user/registration', core_api_views.UserViewSet, base_name='user')
```
we can have the url pattern be the same as the others. also the `base_name` doesn't imply registration.

giddygitau committed:
[Feature #140648421] Changes url to point to the relevant resource

giddygitau committed:
[Feature #140648421] Adds authentication and permission classes
Adds list users functionality

giddygitau committed:
[Feature #140648421] Makes urls more dynamic
One endpoint to ease testing of creation and listing of users

giddygitau committed:
[Feature #140648421] Corrects pep8 line too long warning

giddygitau committed:
[Feature #140648421] Changes the url and the view name to make them more descriptive

giddygitau committed:
[Feature #140648421] Adds email validation and default profile setting during registration via API

giddygitau committed:
[Feature #140648421] Renames test to be more descriptive
Adds url and url_detail to point to the modified endpoint
Adds email to testing

giddygitau committed:
[Feature #140648421] Imports Gym and GymUserConfig
Gym and GymUserConfig are necessary for creating the necessary default gyms

egichuri commented on wger/core/tests/test_register_user_api.py:
```diff
@@ -0,0 +1,16 @@
+    resource = User
+    private_resource = True
+    special_endpoints = ()
+    data = {'username': 'giddy254', 'email': 'giddy@example.com', 'password': 'password'}
```
@giddygitau did you add the pending tests?

giddygitau replied:
The part of the code for creating user profiles was not being run because some imports had not been made, after making the relevant imports, all the code was run by tests.

egichuri commented on wger/core/api/serializers.py:
```diff
@@ -26,6 +28,18 @@
     WeightUnit)
 
 
+class UserRegistrationSerializer(serializers.ModelSerializer):
```
we can have this being more general (`UserSerializer`). In case we need to use it to also list the users

egichuri commented on wger/core/api/views.py:
```diff
@@ -29,14 +32,82 @@
+from rest_framework import serializers
+from wger.config.models import GymConfig
+from wger.gym.models import Gym
+from wger.gym.models import GymUserConfig
```
This is more of a nitpick, but the imports can be arranged better here -- system -> package -> user

giddygitau committed:
[Feature #140648421] Renames UserRegistrationSerializer to UserSerializer for future reusability

giddygitau committed:
[Feature #140648421] Arranges imports in conventional order
